File: sender-mac/Sources/SenderMac/LatencyMessages.swift
import Foundation

struct FrameTimestampMessage: Codable, Sendable {
    let type: String
    let sequence: Int64
    let captureTimeMs: Int64
    let encodeStartTimeMs: Int64
    let encodeEndTimeMs: Int64
    let sendTimeMs: Int64

    init(sequence: Int64, captureTimeMs: Int64, encodeStartTimeMs: Int64, encodeEndTimeMs: Int64, sendTimeMs: Int64) {
        self.type = "frame-timestamp"
        self.sequence = sequence
        self.captureTimeMs = captureTimeMs
        self.encodeStartTimeMs = encodeStartTimeMs
        self.encodeEndTimeMs = encodeEndTimeMs
        self.sendTimeMs = sendTimeMs
    }

    func jsonString() -> String? {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys]
        guard let data = try? encoder.encode(self) else {
            return nil
        }
        return String(data: data, encoding: .utf8)
    }
}

struct FrameLatencyReportMessage: Codable, Sendable {
    let type: String
    let sequence: Int64
    let captureTimeMs: Int64
    let encodeEndTimeMs: Int64
    let receiverDataTimeMs: Int64
    let firstFrameSeenTimeMs: Int64
    let renderSubmitTimeMs: Int64
    let estimatedAppLatencyMs: Int64
}

File: receiver-quest/Assets/Scripts/Latency/LatencyMessages.cs
using System;
using Newtonsoft.Json;

namespace X5Quest.Latency
{
    [Serializable]
    public sealed class FrameTimestampMessage
    {
        public string type;
        public long sequence;
        public long captureTimeMs;
        public long encodeStartTimeMs;
        public long encodeEndTimeMs;
        public long sendTimeMs;

        public static bool TryParse(string json, out FrameTimestampMessage message)
        {
            message = null;
            try
            {
                var parsed = JsonConvert.DeserializeObject<FrameTimestampMessage>(json);
                if (parsed?.type != "frame-timestamp")
                {
                    return false;
                }

                message = parsed;
                return true;
            }
            catch
            {
                return false;
            }
        }
    }

    [Serializable]
    public sealed class FrameLatencyReportMessage
    {
        public string type = "frame-latency-report";
        public long sequence;
        public long captureTimeMs;
        public long encodeEndTimeMs;
        public long receiverDataTimeMs;
        public long firstFrameSeenTimeMs;
        public long renderSubmitTimeMs;
        public long estimatedAppLatencyMs;

        public string ToJson()
        {
            return JsonConvert.SerializeObject(this);
        }
    }

    public sealed class LatencyTracker
    {
        public FrameLatencyReportMessage LastReport { get; private set; }

        public FrameLatencyReportMessage OnFrameTimestamp(FrameTimestampMessage timestamp)
        {
            var now = UnixTimeMs();
            LastReport = new FrameLatencyReportMessage
            {
                sequence = timestamp.sequence,
                captureTimeMs = timestamp.captureTimeMs,
                encodeEndTimeMs = timestamp.encodeEndTimeMs,
                receiverDataTimeMs = now,
                firstFrameSeenTimeMs = now,
                renderSubmitTimeMs = now,
                estimatedAppLatencyMs = now - timestamp.captureTimeMs
            };
            return LastReport;
        }

        public void MarkRenderSubmit()
        {
            if (LastReport == null)
            {
                return;
            }

            LastReport.renderSubmitTimeMs = UnixTimeMs();
            LastReport.estimatedAppLatencyMs = LastReport.renderSubmitTimeMs - LastReport.captureTimeMs;
        }

        private static long UnixTimeMs()
        {
            return DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
        }
    }
}

